libr/core/canal.c

```diff
@@ -5637,46 +5646,59 @@ R_API void r_core_anal_propagate_noreturn(RCore *core, ut64 addr) {
 				eprintf ("Cannot analyze opcode at 0x%08" PFMT64x "\n", xref->addr);
 				continue;
 			}
+			ut64 call_addr = xref->addr;
+			ut64 chop_addr = call_addr + xrefop->size;
 			r_anal_op_free (xrefop);
 			if (xref->type != R_ANAL_REF_TYPE_CALL) {
 				continue;
 			}
-			if (addr != UT64_MAX) {
-				f = r_anal_get_fcn_in (core->anal, xref->addr, 0);
-				if (f) {
-					if (f->addr != addr) {
-						continue;
-					}
-				} else {
-					continue;
-				}
-			} else {
-				f = r_anal_get_fcn_in (core->anal, xref->addr, 0);
-				if (!f || (f->type != R_ANAL_FCN_TYPE_FCN && f->type != R_ANAL_FCN_TYPE_SYM)) {
-					continue;
-				}
-				ut64 addr = f->addr;
-
-				r_anal_fcn_del_locs (core->anal, addr);
-				// big depth results on infinite loops :( but this is a different issue
-				r_core_anal_fcn (core, addr, UT64_MAX, R_ANAL_REF_TYPE_NULL, 3);
 
-				f = r_anal_get_function_at (core->anal, addr);
-				if (!f || (f->type != R_ANAL_FCN_TYPE_FCN && f->type != R_ANAL_FCN_TYPE_SYM)) {
-					continue;
+			// Find the block that has an instruction at exactly the xref addr
+			RList *blocks = r_anal_get_blocks_in (core->anal, call_addr);
+			if (!blocks) {
+				continue;
+			}
+			RAnalBlock *block = NULL;
+			RListIter *bit;
+			RAnalBlock *block_cur;
+			r_list_foreach (blocks, bit, block_cur) {
+				if (r_anal_block_op_starts_at (block_cur, call_addr)) {
+					block = block_cur;
+					break;
 				}
 			}
+			if (block) {
+				r_anal_block_ref (block);
+			}
+			r_list_free (blocks);
+			if (!block) {
+				continue;
+			}
 
-			bool found = false;
-			found = ht_uu_find (done, f->addr, &found);
-			if (f->addr && !found && analyze_noreturn_function (core, f)) {
-				f->is_noreturn = true;
-				r_anal_noreturn_add (core->anal, NULL, f->addr);
-				ut64 *n = malloc (sizeof (ut64));
-				*n = f->addr;
-				r_list_append (todo, n);
-				ht_uu_insert (done, *n, 1);
+			if (request_fcn) {
+				// specific function requested, check if it contains the bb
+				if (!r_list_contains (block->fcns, request_fcn)) {
+					goto kontinue;
+				}
+			} else {
+				r_anal_block_chop_noreturn (block, chop_addr);
+			}
+
+			RListIter *fit;
+			r_list_foreach (block->fcns, fit, f) {
+				bool found = false;
+				found = ht_uu_find (done, f->addr, &found);
+				if (f->addr && !found && analyze_noreturn_function (core, f)) {
+					f->is_noreturn = true;
+					r_anal_noreturn_add (core->anal, NULL, f->addr);
+					ut64 *n = malloc (sizeof (ut64));
```

Comment:
use ut64_new()

Reply:
There is no such thing, just internally in cmd_debug.c

Reply:
just do `HEAPTYPE(ut64);` on top of your file. and you will get this magic ut64_new